src/distr/integer.rs

```diff
@@ -107,21 +107,50 @@ impl_nzint!(NonZeroI64, NonZeroI64::new);
 impl_nzint!(NonZeroI128, NonZeroI128::new);
 
 #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
-macro_rules! x86_intrinsic_impl {
-    ($meta:meta, $($intrinsic:ident),+) => {$(
-        #[cfg($meta)]
-        impl Distribution<$intrinsic> for StandardUniform {
-            #[inline]
-            fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> $intrinsic {
-                // On proper hardware, this should compile to SIMD instructions
-                // Verified on x86 Haswell with __m128i, __m256i
-                let mut buf = [0_u8; core::mem::size_of::<$intrinsic>()];
-                rng.fill_bytes(&mut buf);
-                // x86 is little endian so no need for conversion
-                zerocopy::transmute!(buf)
-            }
-        }
-    )+};
+impl Distribution<__m128i> for StandardUniform {
+    #[inline]
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> __m128i {
+        // NOTE: It's tempting to use the u128 impl here, but confusingly this
+        // results in different code (return via rdx, r10 instead of rax, rdx
+        // with u128 impl) and is much slower (+130 time). This version calls
+        // impls::fill_bytes_via_next but performs well.
+
+        let mut buf = [0_u8; core::mem::size_of::<__m128i>()];
+        rng.fill_bytes(&mut buf);
+        // x86 is little endian so no need for conversion
+
+        // SAFETY: both source and result types are valid for all values.
+        unsafe { core::mem::transmute(buf) }
```

Comment:
I was wondering if here (and elsewhere) we should be using the various unaligned-load intrinsics:
  - [`_mm_loadu_si128`](https://doc.rust-lang.org/core/arch/x86_64/fn._mm_loadu_si128.html)
  - [`_mm256_loadu_si256`](https://doc.rust-lang.org/core/arch/x86_64/fn._mm256_loadu_si256.html) 
  - [`_mm512_loadu_si512`](https://doc.rust-lang.org/core/arch/x86_64/fn._mm512_loadu_si512.html)

instead of a raw `transmute`. However, they seem to produce identical assembly ([AVX2 256-bit Godbolt example](https://godbolt.org/z/fsT8snzsE)), so I guess it's just a matter of taste.

Reply:
`transmute` will ensure `src` and `dest` have the same size and suitable alignment. `transmute_copy` and the load functions above won't check this.

IMO the above (i.e. value semantics) is reason enough to prefer `transmute`.